core/BranchPred.cpp

```diff
@@ -0,0 +1,70 @@
+#include "BranchPred.hpp"
+
+/*
+ * The algorithm used for prediction / update is as follows:
+ * Prediction:
+ *    - look up BHT to determine if the branch is predicted taken or not
+ *    - look up BTB to see if an entry exists for the input fetch pc
+ *       - if present in BTB and predicted taken, BTB entry is used to determine
+ *         prediction branch idx and predictedPC
+ *       - if present in BTB but predicted not taken, BTB entry is used to determine
+ *         prediction branch idx, while predictedPC is the fall through addr
+ *       - if not present in BTB entry, prediction branch idx is the last instr of
+ *         the FetchPacket, while predicted PC is the fall through addr. Also, create
+ *         a new BTB entry 
+ * Update:
+ *    - a valid BTB entry must be present for fetch PC
+ *    - TBD 
+ *
+ */
+
+#define BYTES_PER_INST 4
+
+void SimpleBranchPredictor::updatePredictor(DefaultUpdate & update) {
+
+    sparta_assert(branch_target_buffer_.find(update.FetchPC) != branch_target_buffer_.end());
+    branch_target_buffer_[update.FetchPC].branch_idx = update.branch_idx;
+    if (update.actuallyTaken) {
+        branch_history_table_[update.FetchPC] =
+            (branch_history_table_[update.FetchPC] == 3) ? 3 :
+             branch_history_table_[update.FetchPC] + 1;
+        branch_target_buffer_[update.FetchPC].predictedPC = update.correctedPC;
+    } else {
+        branch_history_table_[update.FetchPC] =
+            (branch_history_table_[update.FetchPC] == 0) ? 0 :
+             branch_history_table_[update.FetchPC] - 1;
+    }
+}
+
+DefaultPrediction SimpleBranchPredictor::getPrediction(const DefaultInput & input) {
+    bool predictTaken;
+    if (branch_history_table_.find(input.FetchPC) != branch_history_table_.end()) {
+        predictTaken = (branch_history_table_[input.FetchPC] > 1);
```

Comment:
A comment noting that history-table states of "3" and "2" as taken vs. "1" and "0" as not-taken would be helpful.  A more detailed header description making note of the implemented algorithm would negate this suggestion to add a comment here.